### ISSUE TITLE
Unbalanced rules for cross checking

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/fmi-standard#274]
**Reported by apillekeit on 15 Jul 2014 12:35 UTC**
**Unbalanced regarding the need for cooperation**

As a follow up from the mail discussion I see an unbalance in the required effort for getting a „green“ FMI compliance mark for importing and exporting tools. See section 2 of the cross check rules.

An exporting tool just needs to export 3 FMUs that pass the FMUCrossChecker and it gets a „green“ FMI compliance mark. Even regardless the FMI variant or platform. That can even be interpreted in a way that the same model exported as FMU for win32, win64 and linux32 would be sufficient.

An importing tool needs to check the import of „all“ FMUs of the supported tool and “all” reference FMUs on the svn not even limited to supported platforms. I cannot find a rule for that, but the current practice is to give a “green” FMI compliance mark if at least 3 FMUs have been imported and simulated successfully (right?). If that rule is not included it should be included in the cross check rules to be in line with the practice.

Since an important intention of FMI is the exchange of FMUs between tools, the rules for importers and exporters should match in a way that the tools are encouraged to work together.

The effort for importing tools to understand, localize and discuss appearing problems with the exporting tools is very high.

Therefore also exporters should be encouraged to cooperate with importing tools to receive a “green” FMI-Compliance mark. E.g. by requiring that at least one importing tool next to the FMUCrossChecker reports the successful import of 3 FMUs with matching simulation results.

**Unbalanced regarding the coverage of use cases **

In general the fact that FMI can be used for different use cases is not covered well by the cross checking rules. Tools that are optimized for certain use cases (e.g. real-time simulation) should also be able to document their FMI compliance. Some prerequisites are needed to prepare models for real-time simulation, that are not related to FMI compliance. Therefore rules like rule number 10 „_…must successfully import all Reference FMUs provided on the SVN server and must successfully import all FMUs…_“ should be limited to the platform, FMI variant and use cases the FMUs are created for and to which the tools claim support for.

**Summary:**

__The proposal is__

 * To require the export of „at least 3 FMUs“ for each supported platform, FMI variant and use case and
 * To require a report of successful import and simulation of "at least 3 FMUs"  by "at least one" importing tool (next to the FMUComplianceChecker) to grant a green FMI-compliance mark for  exporter tools for each supported platform, FMI variant or use case
 * To limit the required number of successfully simulated FMUs to get a "green" FMI-compliance mark for importers to „at least 3 FMUs“ for each supported platform, FMI variant and use case.

----

Migrated-From: https://trac.fmi-standard.org/ticket/274

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#46;junghanns on 16 Jul 2014 11:13 UTC**
> An exporting tool just needs to export 3 FMUs that pass the FMUCrossChecker and it gets a „green“ FMI compliance mark. Even regardless the FMI variant or platform. That can even be interpreted in a way that the same model exported as FMU for win32, win64 and linux32 would be sufficient.

I agree. We will see if the group agrees to tighten those requirements that much.
Many already do what you propose anyway but a significant number does not.

Proposal:
Now you have to export 3 FMUs for each supported FMI Variant and supported platform.

I fixed this in the Rules document. That needs fixing in the script once v3 is approved. 

> An importing tool needs to check the import of „all“ FMUs of the supported tool and “all” reference FMUs on the svn not even limited to supported platforms. I cannot find a rule for that, but the current practice is to give a “green” FMI compliance mark if at least 3 FMUs have been imported and simulated successfully (right?). If that rule is not included it should be included in the cross check rules to be in line with the practice.

Well, you need to check and report - that is different from passing those.
This was meant to avoid people from reporting only on the FMUs they pass.
But with FMUs added later, this might not work as intended - or the script has to do even more complex acrobatic using time on top what is already done now.
Not to mention how complicated the rules would be.

Should we add "untested" to the current "passed/rejected/failed" counts?

> Since an important intention of FMI is the exchange of FMUs between tools, the rules for importers and exporters should match in a way that the tools are encouraged to work together.
> 
> The effort for importing tools to understand, localize and discuss appearing problems with the exporting tools is very high.

But that is symmetrical, right? Such a conversation requires two.
But your next point is really important - exporters that might not cooperate with importers to get problems fixed because they are already "green":

> Therefore also exporters should be encouraged to cooperate with importing tools to receive a “green” FMI-Compliance mark. E.g. by requiring that at least one importing tool next to the FMUCrossChecker reports the successful import of 3 FMUs with matching simulation results.

Therefore I changed the text to require "at least one importing tool has to report successful import of at least 3 FMUs per supported FMI Variant and supported platform."

> **Unbalanced regarding the coverage of use cases **
> 
> In general the fact that FMI can be used for different use cases is not covered well by the cross checking rules. Tools that are optimized for certain use cases (e.g. real-time simulation) should also be able to document their FMI compliance. Some prerequisites are needed to prepare models for real-time simulation, that are not related to FMI compliance. Therefore rules like rule number 10 „_…must successfully import all Reference FMUs provided on the SVN server and must successfully import all FMUs…_“ should be limited to the platform, FMI variant and use cases the FMUs are created for and to which the tools claim support for.

You are not alone here. Others are struggling with optimization, or whatever. So, real-time applications should not be singled out.
But you point is valid, for the more general case: With all the capability flags we now have, what if an importer requires some FMU capability that is not present in the reference FMU(s)?
So, I softened Rule 10 significantly - please le me know what you think:
```
FMU importing tools must report on importing for all Reference FMUs available for the supported FMI Variant and supported platforms provided on the SVN server (Note: not required to “pass” because of potentially missing capabilities of the FMU) and must successfully import at least 3 FMUs for each supported FMI Variant and supported platform of at least 3 exporting tools, and run for as long as the {FMUName]_ref.opt states and supply a .csv file of the solution they computed.
```

Note: I am not sure what you mean with use-case. If you mean what I describe as "capability" (as the standard does) - than I think I answered fully here. If not, please explain again.

**Comment by modelica-trac-importer** [comment]
**Comment by apillekeit on 16 Jul 2014 14:28 UTC**
Replying to [comment:1 andreas.junghanns]:


Now the rules of section 5 and the rules described in section 2 are somewhat inconsistent regarding the requirements for getting a "green" mark for exporters and importers. 

I don't fully understand the structure of the cross check document. 

If it is intended that section 5 is derived from the previous sections, then the rules should be written in a consistent way. 

If all the rules of section 5 and the previous sections together represent the cross check rules, I guess it is hard to understand but ok now.


> Should we add "untested" to the current "passed/rejected/failed" counts?

I think untested FMUs should not be treated as passed, rejected or failed. Since these FMUs are simply not tested yet. But I also think the website will start to become confusing if too much information is displayed.




> > You are not alone here. Others are struggling with optimization, or whatever. So, real-time applications should not be singled out.

I guess it is ok if I reply that friendly remark with a friendly: No, you are not alone here.

The FMI project was quite centered on Modelica and what I would call “non-real-time” tools. The traditional members of the project start to understand that new members and new use cases (and I really do not only talk about real-time here) may have special requirements that need to be taken care for in a way that it is not impossible to comply with FMI rules. But I think the project is on the right track. Thank you for listening and for adapting the rules.

**Comment by modelica-trac-importer** [comment]
**Comment by andreas&#46;junghanns on 16 Jul 2014 15:22 UTC**
Replying to [comment:2 apillekeit]:
> Replying to [comment:1 andreas.junghanns]:
> Now the rules of section 5 and the rules described in section 2 are somewhat inconsistent regarding the requirements for getting a "green" mark for exporters and importers. 
> I don't fully understand the structure of the cross check document. 

Good point. This is not clear (any more?).
The intend was to have section 2 to explain roughly the UI and meaning.
And then have the rules in section 5 define it in detail.
I messed that up - thanks for pointing this out!
I will give this another round to see if I can make the intent of section 2 clear and at the same time removing details there that are supposed to be defined in the rules.
This way, we will have less chance for conflicts and the reader can understand more easily.
Please check my latest changes.
 
> > Should we add "untested" to the current "passed/rejected/failed" counts?
> 
> I think untested FMUs should not be treated as passed, rejected or failed. Since these FMUs are simply not tested yet. But I also think the website will start to become confusing if too much information is displayed.

I know. For now we can simply ignore this by leaving things as is.
Once we see abuse, we will have to step in and deal with it,
even if it means to complicate the information presented.

> > > You are not alone here. Others are struggling with optimization, or whatever. So, real-time applications should not be singled out.
> 
> I guess it is ok if I reply that friendly remark with a friendly: No, you are not alone here.
> 
> The FMI project was quite centered on Modelica and what I would call “non-real-time” tools. The traditional members of the project start to understand that new members and new use cases (and I really do not only talk about real-time here) may have special requirements that need to be taken care for in a way that it is not impossible to comply with FMI rules. But I think the project is on the right track. Thank you for listening and for adapting the rules.

Well, this will have to be discussed in the Steering Committee and with the Advisory Board:
How far do we stretch to go beyond what is currently possible?
At some point we have to draw the line in order to ensure the continued success of the standard.

If we want too much, we might not be able to do even the little we do now with reasonable effort.
I am glad technicians run this now, as a sense for complexity seems to still drive most decisions.

**Comment by modelica-trac-importer** [comment]
**Modified by andreas&#46;junghanns on 16 Jul 2014 21:37 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by andreas&#x2e;junghanns on 16 Jul 2014 22:05 UTC**
